File: code/0015. 3Sum/solution2.cpp
class Solution {
private:
    unordered_set<int> visited;
    
    void twoSum(vector<int>& nums, int target, int start, vector<vector<int>>& result) {
        int n = nums.size();
        
        for (int i = start; i < n; i++) {
            int num = nums[i];
            if (visited.count(-target - num) > 0) {
                result.push_back({target, -target - num, num});
                while (i + 1 < n && nums[i + 1] == num) {
                    i++;
                }
            }
            else {
                visited.insert(num);
            }
        }
        
        visited.clear();
    }
    
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());
        vector<vector<int>> result;
        
        for (int i = 0; i < n; i++) {
            if (i == 0 || nums[i] != nums[i - 1]) {
                int target = nums[i];
                twoSum(nums, target, i + 1, result);
            }
        }
        
        return result;
    }
};

File: code/0259. 3Sum Smaller/solution2.java
class Solution {
    public int threeSumSmaller(int[] nums, int target) {
        int n = nums.length, result = 0;
        Arrays.sort(nums);
        
        for (int i = 0; i < n; i++) {
            int num = nums[i];
            int p1 = i + 1, p2 = n - 1;
            while (p1 < p2) {
                if (num + nums[p1] + nums[p2] < target) {
                    result += p2 - p1;
                    p1++;
                }
                else {
                    p2--;
                }
            }
        }
        
        return result;
    }
}